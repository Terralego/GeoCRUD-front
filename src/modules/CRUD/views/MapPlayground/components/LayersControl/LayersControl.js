import React from 'react';
import PropTypes from 'prop-types';
import { Icon, Popover, Checkbox, Position } from '@blueprintjs/core';
import Api from '@terralego/core/modules/Api';

import AbstractMapControl from '@terralego/core/modules/Map/helpers/AbstractMapControl';
import Tooltip from '@terralego/core/components/Tooltip';
import { toast } from '../../../../../../utils/toast';

import './styles.scss';

export class LayersControl extends AbstractMapControl {
  static containerClassName = 'mapboxgl-ctrl mapboxgl-ctrl-group mapboxgl-ctrl-layers';

  static propTypes = {
    layers: PropTypes.arrayOf(PropTypes.shape({})),
    relations: PropTypes.arrayOf(PropTypes.shape({})),
    loadSourceAndLayerById: PropTypes.func,
    getMapStyle: PropTypes.func,
  };

  static defaultProps = {
    layers: [],
    relations: [],
    getMapStyle() {},
    loadSourceAndLayerById() {},
    onChange() {},
  };

  state = {
    layers: [],
  };

  componentDidMount() {
    const { layers = [], map } = this.props;
    layers.forEach(({ id }) => {
      if (map.getSource(id)) {
        this.setLayoutProperty(id, map.getLayoutProperty(id, 'visibility') === 'visible');
      }
    });
    this.manageRelationsLayers();
  }

  componentDidMUpdate({ relations: prevRelations }) {
    const { relations } = this.props;
    if (prevRelations !== relations) {
      this.manageRelationsLayers();
    }
  }

  componentWillUnmount() {
    const { map, relations } = this.props;
    relations.forEach(({ crud_view_pk: id }) => {
      const layerId = this.getRelationLayerID(id);
      if (map.getSource(layerId)) {
        map.setLayoutProperty(layerId, 'visibility', 'none');
      }
    });
  }

  manageRelationsLayers = () => {
    const { layers, relations } = this.props;
    const nextLayers = layers.map(layer => {
      if (layer.view_source !== 'relation') {
        return layer;
      }
      return {
        ...layer,
        ...relations.find(({ id_layer_vt: idLayerVectorTiles } = {}) => idLayerVectorTiles === layer['source-layer'])
      };
    });
    this.setState({ layers: nextLayers });
  };

  getAllRelationData = async ({ method = 'GET', endpoint, body, querystring }) => {
    const { page = 1, pageSize = 100, ...qs } = querystring || {};

    // Get first page to know total
    const { results: firstPage, count } = await Api.request(endpoint, {
      method,
      body,
      querystring: { page, page_size: pageSize, ...qs },
    });

    const pageCount = Math.ceil(count / pageSize);

    const promises = Array.from({ length: pageCount - 1 }, async (_, index) => {
      const { results: nextPage } = await Api.request(endpoint, {
        method,
        body,
        querystring: { page: index + 2, page_size: pageSize, ...qs },
      });
      return nextPage;
    });

    return [firstPage, ...(await Promise.all(promises))].flat().map(({ identifier }) => identifier);
  };

  onChange = async ({ target: { value, checked } }, url) => {
    const { map } = this.props;
    if (url) {
      let data = {};
      try {
        const request = await this.getAllRelationData({ endpoint: url.replace('/api/', '') });
        data = request;
      } catch (e) {
        toast.displayError(e.message);
        return;
      }
      map.setFilter(value, ['in', '_id', ...data]);
      this.setState(({ layers: prevLayers }) => ({
        layers: prevLayers.map(layer =>
          layer.url !== url
            ? layer
            : // To load data once
              { ...layer, url: null },
        ),
      }));
    }
    this.setLayoutProperty(value, checked);
  };

  getRelationLayerID = id => {
    const { featureID } = this.props;
    return `CRUD-${featureID}-relation-${id}`;
  };

  setLayoutProperty = async (layerId, isVisible) => {
    const { map, layers, loadSourceAndLayerById } = this.props;
    const { source } = layers.find(layer => layer.id === layerId);
    const sourceId = Number(source);
    if (map.getSource(sourceId)) {
      map.setLayoutProperty(layerId, 'visibility', isVisible ? 'visible' : 'none');
    } else if (!Number.isNaN(sourceId)) {
      await loadSourceAndLayerById(sourceId);
      map.setLayoutProperty(layerId, 'visibility', isVisible ? 'visible' : 'none');
    }
  };

  onClosing = () => {
    this.forceUpdate();
  };

  render() {
    const { map, translate } = this.props;
    const { layers } = this.state;

    if (!layers.length) {
      return null;
    }

    const extraGeomLayers = layers.filter(layer => layer.view_source !== 'relation');
    const relationsLayers = layers.filter(layer => layer.view_source === 'relation');

    return (
      <Tooltip content={translate('CRUD.map.controls.layers.label')} className="layerGroup">
        <Popover
          className="popoverPos"
          onClosing={this.onClosing}
          position={Position.BOTTOM_LEFT}
          content={
            <div className="radioGroup layerGroup__item">
              {extraGeomLayers.length > 0 && (
                <>
                  <h3 className="layerGroup__title">
                    {translate('CRUD.map.controls.layers.title')}
                  </h3>
                  {extraGeomLayers.map(({ title, id, empty = false, source }) => {
                    const defaultChecked =
                      !empty &&
                      map.getSource(source) &&
                      map.getLayoutProperty(id, 'visibility') === 'visible';
                    return (
                      <Checkbox
                        className="bgLayer-radio"
                        defaultChecked={defaultChecked}
                        disabled={empty}
                        key={id}
                        onChange={this.onChange}
                        label={title}
                        value={id}
                      />
                    );
                  })}
                </>
              )}
              {relationsLayers.length > 0 && (
                <>
                  <h3 className="layerGroup__title">
                    {translate('CRUD.map.controls.relations.title')}
                  </h3>
                  {relationsLayers.map(({ title, id, empty = false, source, url }) => (
                    <Checkbox
                      className="bgLayer-radio"
                      defaultChecked={
                        !url &&
                        map.getSource(source) &&
                        map.getLayoutProperty(id, 'visibility') === 'visible'
                      }
                      disabled={empty}
                      key={id}
                      onChange={event => this.onChange(event, url)}
                      label={title}
                      value={id}
                    />
                  ))}
                </>
              )}
            </div>
          }
        >
          <button
            className="mapboxgl-ctrl-icon"
            type="button"
            aria-label={translate('CRUD.map.controls.layers.label')}
          >
            <Icon icon="multi-select" />
          </button>
        </Popover>
      </Tooltip>
    );
  }
}

export default LayersControl;
