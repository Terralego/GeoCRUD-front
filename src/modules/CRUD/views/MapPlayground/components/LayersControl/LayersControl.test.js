import React from 'react';
import renderer from 'react-test-renderer';
import LayersControl from './LayersControl';

jest.mock('../../../../../../utils/toast', () => ({
  toast: {
    displayError: jest.fn(),
  },
}));

jest.mock('@blueprintjs/core', () => ({
  Checkbox: props => <input type="checkbox" {...props} />,
  Icon: props => <div {...props} />,
  Popover: props => <div {...props} />,
  Position: {
    BOTTOM_LEFT: 'bottom-left',
  },
}));

jest.mock('@terralego/core/modules/Map/InteractiveMap', () => {
  const InteractiveMap = () => <div>InteractiveMap</div>;
  InteractiveMap.INTERACTION_FN = 'function';
  return InteractiveMap;
});

jest.mock('@terralego/core/modules/Map', () => ({}));

jest.mock('@terralego/core/components/Tooltip', () => props => <div {...props} />);

const defaultProps = {
  translate: text => text,
  map: {
    addLayer: jest.fn(),
    addSource: jest.fn(),
    setLayoutProperty: jest.fn(),
    getSource: id => id === 'foo',
    getLayoutProperty: id => (id === 'foo' ? 'visible' : 'none'),
  },
};

describe('snapshots', () => {
  it('should not display anything', () => {
    const tree = renderer
      .create(<LayersControl layers={[]} relations={[]} {...defaultProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should display layers', () => {
    const tree = renderer
      .create(
        <LayersControl
          {...defaultProps}
          layers={[
            {
              id: 'id1',
              'source-layer': 'foo',
              source: '1',
              title: 'Foo',
            },
            {
              id: 'id2',
              'source-layer': 'bar',
              source: '2',
              title: 'Bar',
            },
          ]}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should display relations', () => {
    const tree = renderer
      .create(
        <LayersControl
          {...defaultProps}
          layers={[
            {
              id: 'id1',
              'source-layer': 'foo',
              source: '1',
              title: 'POIs',
              view_source: 'relation',
            },
            {
              id: 'id2',
              'source-layer': 'bar',
              source: '2',
              title: 'Touristics content',
              view_source: 'relation',
            },
          ]}
          relations={[
            {
              label: 'POIs',
              order: 40,
              url: '/api/crud/layers/29/features/foo/relation/11/features/',
              geojson: '/api/crud/layers/29/features/foo/relation/11/features.geojson',
              crud_view_pk: 17,
              id_layer_vt: 'foo',
              empty: false,
            },
            {
              label: 'Touristics content',
              order: 1,
              url: '/api/crud/layers/29/features/bar/relation/11/features/',
              geojson: '/api/crud/layers/29/features/bar/relation/11/features.geojson',
              crud_view_pk: 2,
              id_layer_vt: 'bar',
              empty: true,
            },
          ]}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
