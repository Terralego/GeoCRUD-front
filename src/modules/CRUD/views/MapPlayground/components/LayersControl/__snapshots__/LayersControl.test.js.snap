// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should display layers 1`] = `
<div
  className="layerGroup"
  content="CRUD.map.controls.layers.label"
>
  <div
    className="popoverPos"
    content={
      <div
        className="radioGroup layerGroup__item"
      >
        <React.Fragment>
          <h3
            className="layerGroup__title"
          >
            CRUD.map.controls.layers.title
          </h3>
          <Checkbox
            className="bgLayer-radio"
            defaultChecked={false}
            disabled={false}
            label="Foo"
            onChange={[Function]}
            value="id1"
          />
          <Checkbox
            className="bgLayer-radio"
            defaultChecked={false}
            disabled={false}
            label="Bar"
            onChange={[Function]}
            value="id2"
          />
        </React.Fragment>
      </div>
    }
    onClosing={[Function]}
    position="bottom-left"
  >
    <button
      aria-label="CRUD.map.controls.layers.label"
      className="mapboxgl-ctrl-icon"
      type="button"
    >
      <div
        icon="multi-select"
      />
    </button>
  </div>
</div>
`;

exports[`snapshots should display relations 1`] = `
<div
  className="layerGroup"
  content="CRUD.map.controls.layers.label"
>
  <div
    className="popoverPos"
    content={
      <div
        className="radioGroup layerGroup__item"
      >
        <React.Fragment>
          <h3
            className="layerGroup__title"
          >
            CRUD.map.controls.relations.title
          </h3>
          <Checkbox
            className="bgLayer-radio"
            defaultChecked={false}
            disabled={false}
            label="POIs"
            onChange={[Function]}
            value="id1"
          />
          <Checkbox
            className="bgLayer-radio"
            defaultChecked={false}
            disabled={true}
            label="Touristics content"
            onChange={[Function]}
            value="id2"
          />
        </React.Fragment>
      </div>
    }
    onClosing={[Function]}
    position="bottom-left"
  >
    <button
      aria-label="CRUD.map.controls.layers.label"
      className="mapboxgl-ctrl-icon"
      type="button"
    >
      <div
        icon="multi-select"
      />
    </button>
  </div>
</div>
`;

exports[`snapshots should not display anything 1`] = `null`;
